File: examples/python/11_transactions.py
import redis

r = redis.Redis(host='localhost', port=6379, decode_responses=True)

print("--- 11 Transactions (MULTI/EXEC) ---")

key1 = "account:A"
key2 = "account:B"

r.set(key1, 100)
r.set(key2, 50)

print(f"Initial: A={r.get(key1)}, B={r.get(key2)}")

# Start Transaction
print("Starting Transaction...")
pipe = r.pipeline(transaction=True) # transaction=True uses MULTI/EXEC

pipe.decrby(key1, 20)
pipe.incrby(key2, 20)

# Execute
print("Executing...")
results = pipe.execute()

print(f"Results: {results}") # [80, 70]
print(f"Final: A={r.get(key1)}, B={r.get(key2)}")

# Discard Example
print("\n[DISCARD Test]")
pipe = r.pipeline(transaction=True)
pipe.set("temp_key", "should_not_exist")
print("Discarding transaction...")
# redis-py's pipeline wraps DISCARD logic in reset(), but strictly speaking we interact via commands if doing manually.
# Using raw commands to demonstrate:
try:
    r.execute_command("MULTI")
    r.execute_command("SET", "temp_key", "fail")
    r.execute_command("DISCARD")
    print("Transaction Discarded.")
except Exception as e:
    print(e)

print(f"temp_key exists? {r.exists('temp_key')}")

File: examples/nodejs/11_transactions.js
const Redis = require("ioredis");
const redis = new Redis({ host: "localhost", port: 6379 });

async function run() {
    console.log("--- 11 Transactions ---");

    const key1 = "node:tx:1";
    const key2 = "node:tx:2";

    await redis.psetex(key1, 10000, "100");
    await redis.psetex(key2, 10000, "200");

    console.log("Beginning Transaction...");
    const pipeline = redis.multi();

    pipeline.incrby(key1, 10);
    pipeline.decrby(key2, 10);

    // Note: ioredis 'multi' abstraction handles MULTI/EXEC automatically when .exec() is called
    const results = await pipeline.exec();

    console.log("Transaction Results:");
    // Results is an array of [error, result] tuples
    results.forEach((res, idx) => {
        console.log(`Cmd ${idx + 1}: Error=${res[0]}, Value=${res[1]}`);
    });

    const v1 = await redis.get(key1);
    const v2 = await redis.get(key2);
    console.log(`Final Values: ${v1}, ${v2}`);

    redis.disconnect();
}

run().catch(console.error);
